### PR TITLE
Integrate with `ferveo` crate

### LINKED ISSUE #49: Integrate and re-export ferveo for public API
Presently all ferveo-generated objects are handled as bytes.  Instead, ferveo types can be implemented directly in `dks.rs` as well as the python/wasm bindings.  This implies that ferveo provides a crate exporting it's python and wasm bindings for consumption in nucypher-core (https://github.com/nucypher/ferveo/issues/87).  

##### targets

- `ThresholdDecryptionRequest.ciphertext` -> `ferveo.api.Ciphertext`
- `ThresholdDecryptionResponse.decryption_share`-> `ferveo.tpke.api.DecryptionShareSimple||DecryptionSharePrecomputed`
- `NodeMetadataPayload.ferveo_public_key` -> `ferveo.common.keypair.PublicKey`

### PR DESCRIPTION
**Pull request by piotr-roslaniec**
- Closes #49 
- Changes in `ferveo` are tracked here: https://github.com/nucypher/ferveo/pull/119
- Re-applies the WASM build fix from: https://github.com/nucypher/nucypher-core/pull/52#discussion_r1193016251
- Questions
  - Should we rename `PublicKey` to `UmbralPublicKey` and `FerveoPublicKey` respectively? Technically, it's sufficient to rename one of them, `FerveoPublicKey`.
  - For now, renaming to `FerveoPublicKey`

### CODE REVIEW

**Comment by fjarri** [comment]
Is it possible to make submodules somehow? Is it a thing in JS?

**Comment by piotr-roslaniec** [comment]
Yes, you can implement modules in JS. For example, you can group your exports like this:
```
// my-module.js
export { MyClass, myFunction };

// run.js
import myModule from 'myModule'
myModule.myFunction();
const { MyClass } = myModule;
``` 

I'm not sure how to do it in `nucypher-core-wasm` yet, but I will give it a go.

**Comment by piotr-roslaniec** [comment]
halp

**Comment by piotr-roslaniec** [comment]
Ok, I just realized this is a known issue:
https://github.com/fjarri/wasm-bindgen-derive/blame/955ed47bef0806fc5eb08a9d011f4bce10b7cf16/wasm-bindgen-derive/src/lib.rs#L269

Should be fixed by updating `wasm-bindgen-derive` and `wasm-bindgen` consistently across all Rust crates.

**Comment by piotr-roslaniec** [comment]
Hi @derekpierre, just wanted to check in with you. Are there any other TODOs that should be covered here?

**Comment by derekpierre** [comment]
> Hi @derekpierre, just wanted to check in with you. Are there any other TODOs that should be covered here?

Perhaps this one as well? https://github.com/piotr-roslaniec/nucypher-core/blob/ferveo-types/nucypher-core/src/lib.rs#L51. Also adding @KPrasch as well as a reviewer.

--

Do these changes need an analogous update on the `nucypher/nucypher` python side of things?

File: nucypher-core-python/src/lib.rs [hunk 1]
```diff
@@ -647,7 +649,7 @@ impl ThresholdDecryptionRequest {
     pub fn new(
         ritual_id: u16,
         variant: u8,
-        ciphertext: &[u8], // TODO use ferveo Ciphertext type
+        ciphertext: &Ciphertext,
```
Comment (KPrasch) [inline nucypher-core-python/src/lib.rs:652]: This change is supremely satisfying to see!

**Review by derekpierre — APPROVED** [review]
🎸

**Pull request merged**